runtime/stdlib/random.go

```diff
@@ -195,3 +205,207 @@ func NewRevertibleRandomFunction(generator RandomGenerator) StandardLibraryValue
+		getRandomBytes(buffer[bufferSize-byteSize:], generator)
+		// big endianness must be used in this case
+		random = binary.BigEndian.Uint64(buffer[:])
+		// adjust to the size of max in bits
+		random &= mask
+	}
+	return random
+}
+
+// cases of a random number of size larger than 8 bytes can be all treated
+// by the same function, based on the big.Int type.
+func getBigRandomNumber(
+	generator RandomGenerator,
+	ty sema.Type,
+	moduloArg interpreter.Value,
+) *big.Int {
+
+	// buffer to get random bytes from the generator
+	// 32 is the size of the largest type supported
+	const bufferSize = 32
+	var buffer [bufferSize]byte
+	// case where no modulo argument was provided
+	if moduloArg == nil {
+		numericType, ok := ty.(*sema.NumericType)
+		if !ok {
+			// checker should prevent this
+			panic(errors.NewUnreachableError())
+		}
+		bytes := numericType.ByteSize()
+		getRandomBytes(buffer[:bytes], generator)
+		// SetBytes considers big endianness (although little endian could be used too)
+		return new(big.Int).SetBytes(buffer[:bytes])
+	}
+
+	var ok bool
+	var modulo *big.Int
+
+	switch ty {
+	case sema.UInt128Type:
+		var moduloVal interpreter.UInt128Value
+		moduloVal, ok = moduloArg.(interpreter.UInt128Value)
+		modulo = moduloVal.BigInt
+	case sema.UInt256Type:
+		var moduloVal interpreter.UInt256Value
+		moduloVal, ok = moduloArg.(interpreter.UInt256Value)
+		modulo = moduloVal.BigInt
+	case sema.Word128Type:
+		var moduloVal interpreter.Word128Value
+		moduloVal, ok = moduloArg.(interpreter.Word128Value)
+		modulo = moduloVal.BigInt
+	case sema.Word256Type:
+		var moduloVal interpreter.Word256Value
+		moduloVal, ok = moduloArg.(interpreter.Word256Value)
+		modulo = moduloVal.BigInt
+	default:
+		// sanity check: shouldn't reach here
+		panic(errors.NewUnreachableError())
+	}
+
+	if !ok {
+		// checker should prevent this
+		panic(errors.NewUnreachableError())
+	}
+
+	// user error if modulo is zero
+	if modulo.Sign() == 0 {
+		panic(ZeroModuloError)
+	}
+
+	// `max` is the maximum value that can be returned (modulo - 1)
+	one := big.NewInt(1)
+	max := new(big.Int).Sub(modulo, one)
+	// count the byte size of `max`
+	bitSize := max.BitLen()
+	byteSize := (bitSize + 7) >> 3
+	// get a bit mask (0b11..11) that covers all `max`'s bits:
+	// `mask` can be computed as:   (1 << bitSize) -1
+	mask := new(big.Int).Lsh(one, uint(bitSize))
+	mask.Sub(mask, one)
+
+	// Generate a number less or equal than `max`
+	// use the reject-sample method to avoid the modulo bias.
+	// the function isn't constant-time in this case and may take longer than computing
+	// a modular reduction.
+	// However, sampling exactly the size of `max` in bits makes the loop return fast:
+	// loop returns after (k) iterations with a probability of at most 1-(1/2)^k.
+	//
+	// (a different approach would be to pull 128 bits more bits than the size of `max`
+	// from the random generator and use big number reduction by `modulo`)
+	random := new(big.Int).Set(modulo)
```

Comment:
This `Set` call likely instantiates a new slice (I think every set call for `big.Int` does) and to me it is confusing why we do it. 

So same suggestion as my previous comment to change the for loop to a posterior break. 

```golang
for {
   ... 
   if random.Cmp(max) <= 0 {
      break
   }
}
```